File: app/lib/src/crypto/safety_code.dart
import 'dart:convert';
import 'dart:typed_data';

import 'package:crypto/crypto.dart' as c;

/// Safety codes let two people verify, out of band, that no server-injected
/// man-in-the-middle has substituted keys. Both devices independently derive the
/// SAME code from BOTH parties' X25519 public keys and compare it in person.
///
/// This is a byte-for-byte mirror of the Go implementation
/// (server/internal/crypto/safetycode.go), canonical scheme "aul-safety-code:v1":
///
///   low, high = sort(pubA, pubB)                       // lexicographic bytes
///   digest    = SHA256("aul-safety-code:v1" || 0x00 || low || high)
///   emoji[i]  = alphabet[digest[i] % 64]               // for i in 0..length
///
/// SHA-256 is used (not a libsodium-only primitive) precisely so the fingerprint
/// is trivially identical across Go, Dart and JS. It is validated against the
/// committed cross-language vectors in /vectors/crypto-vectors.json.
class SafetyCode {
  SafetyCode(this.emojis, this.digest);

  final List<String> emojis;
  final Uint8List digest;

  static const int publicKeyLength = 32;
  static const int length = 10;
  static const String _domain = 'aul-safety-code:v1';

  /// Canonical 64-emoji alphabet (index 0..63), frozen for v1. Must match the
  /// Go `SafetyEmojiAlphabet` exactly.
  static const List<String> alphabet = [
    '🐶', '🐱', '🐭', '🐹', '🐰', '🦊', '🐻', '🐼', //
    '🐨', '🐯', '🦁', '🐮', '🐷', '🐸', '🐵', '🐔',
    '🐧', '🐦', '🦆', '🦉', '🐴', '🦄', '🐝', '🐛',
    '🦋', '🐌', '🐞', '🐢', '🐍', '🐙', '🐠', '🐬',
    '🐳', '🐘', '🐫', '🐑', '🍎', '🍊', '🍋', '🍌',
    '🍉', '🍇', '🍓', '🍒', '🍑', '🍅', '🌽', '🥕',
    '🍔', '🍕', '🍩', '🍪', '🎂', '🍫', '🍭', '🌵',
    '🌲', '🌸', '🌻', '🌈', '🌙', '🔥', '🌊', '💧',
  ];

  /// Derives the canonical safety code from two X25519 public keys. The result
  /// is independent of argument order.
  static SafetyCode compute(Uint8List pubA, Uint8List pubB) {
    if (pubA.length != publicKeyLength || pubB.length != publicKeyLength) {
      throw ArgumentError('public key must be $publicKeyLength bytes');
    }
    var low = pubA;
    var high = pubB;
    if (_compareBytes(pubA, pubB) > 0) {
      low = pubB;
      high = pubA;
    }
    final input = <int>[...ascii.encode(_domain), 0x00, ...low, ...high];
    final digest = Uint8List.fromList(c.sha256.convert(input).bytes);

    final emojis = <String>[
      for (var i = 0; i < length; i++) alphabet[digest[i] % 64],
    ];
    return SafetyCode(emojis, digest);
  }

  /// Emoji code, space-separated.
  String get display => emojis.join(' ');

  /// First 8 digest bytes as grouped hex (accessible alternative to emoji).
  String get hexFallback {
    final h = _hex(digest.sublist(0, 8));
    final parts = <String>[];
    for (var i = 0; i < h.length; i += 4) {
      parts.add(h.substring(i, i + 4));
    }
    return parts.join('-');
  }
}

int _compareBytes(Uint8List a, Uint8List b) {
  final n = a.length < b.length ? a.length : b.length;
  for (var i = 0; i < n; i++) {
    if (a[i] != b[i]) return a[i] - b[i];
  }
  return a.length - b.length;
}

String _hex(List<int> bytes) {
  const digits = '0123456789abcdef';
  final sb = StringBuffer();
  for (final b in bytes) {
    sb.write(digits[(b >> 4) & 0xf]);
    sb.write(digits[b & 0xf]);
  }
  return sb.toString();
}

File: server/internal/crypto/safetycode.go
package crypto

import (
	"bytes"
	"crypto/sha256"
	"encoding/hex"
	"errors"
	"strings"
)

// Safety codes let two people verify, out of band, that no server-injected
// man-in-the-middle has substituted keys. Both devices independently derive the
// SAME code from BOTH parties' X25519 public keys and compare it in person.
//
// Canonical scheme "aul-safety-code:v1" — MUST be mirrored byte-for-byte by the
// Dart and JS clients (a cross-language test vector is checked in):
//
//	low, high := sort(pubA, pubB)                       // lexicographic byte order
//	digest    := SHA256("aul-safety-code:v1" || 0x00 || low || high)
//	emoji[i]  := SafetyEmojiAlphabet[ digest[i] % 64 ]  // for i in 0..SafetyCodeLength
//
// Because 256 is an exact multiple of 64, `digest[i] % 64` is unbiased, giving
// 6 bits per emoji. The default 10-emoji code carries 60 bits: forging a
// matching code requires ~2^60 X25519 keygens (documented in THREAT_MODEL.md as
// the v1 bound). The full digest and a hex fallback are also returned for
// accessibility.
const (
	// PublicKeyLength is the X25519 public key size in bytes.
	PublicKeyLength = 32
	// SafetyCodeLength is the number of emoji in a code.
	SafetyCodeLength = 10
	// safetyCodeDomain domain-separates the fingerprint hash.
	safetyCodeDomain = "aul-safety-code:v1"
)

// ErrInvalidPublicKey is returned for a wrong-length X25519 public key.
var ErrInvalidPublicKey = errors.New("crypto: public key must be 32 bytes")

// SafetyEmojiAlphabet is the canonical 64-emoji alphabet (index 0..63). Each is
// a single Unicode scalar with default emoji presentation (no variation
// selectors or ZWJ), chosen to render consistently across platforms. This list
// is frozen for v1; changing it changes every safety code.
var SafetyEmojiAlphabet = [64]string{
	"🐶", "🐱", "🐭", "🐹", "🐰", "🦊", "🐻", "🐼",
	"🐨", "🐯", "🦁", "🐮", "🐷", "🐸", "🐵", "🐔",
	"🐧", "🐦", "🦆", "🦉", "🐴", "🦄", "🐝", "🐛",
	"🦋", "🐌", "🐞", "🐢", "🐍", "🐙", "🐠", "🐬",
	"🐳", "🐘", "🐫", "🐑", "🍎", "🍊", "🍋", "🍌",
	"🍉", "🍇", "🍓", "🍒", "🍑", "🍅", "🌽", "🥕",
	"🍔", "🍕", "🍩", "🍪", "🎂", "🍫", "🍭", "🌵",
	"🌲", "🌸", "🌻", "🌈", "🌙", "🔥", "🌊", "💧",
}

// SafetyCode is a verifiable device fingerprint derived from two public keys.
type SafetyCode struct {
	Emojis []string // SafetyCodeLength emoji, e.g. ["🐶","🍎",...]
	Digest []byte   // full 32-byte SHA-256 digest
}

// String renders the emoji code space-separated.
func (s SafetyCode) String() string { return strings.Join(s.Emojis, " ") }

// HexFallback returns the first 8 digest bytes as grouped hex (an accessible
// alternative to emoji for users who cannot distinguish them).
func (s SafetyCode) HexFallback() string {
	h := hex.EncodeToString(s.Digest[:8])
	var b strings.Builder
	for i := 0; i < len(h); i += 4 {
		if i > 0 {
			b.WriteByte('-')
		}
		b.WriteString(h[i : i+4])
	}
	return b.String()
}

// ComputeSafetyCode derives the canonical safety code from two X25519 public
// keys. The result is independent of argument order.
func ComputeSafetyCode(pubA, pubB []byte) (SafetyCode, error) {
	if len(pubA) != PublicKeyLength || len(pubB) != PublicKeyLength {
		return SafetyCode{}, ErrInvalidPublicKey
	}
	low, high := pubA, pubB
	if bytes.Compare(pubA, pubB) > 0 {
		low, high = pubB, pubA
	}
	h := sha256.New()
	h.Write([]byte(safetyCodeDomain))
	h.Write([]byte{0x00})
	h.Write(low)
	h.Write(high)
	digest := h.Sum(nil)

	emojis := make([]string, SafetyCodeLength)
	for i := 0; i < SafetyCodeLength; i++ {
		emojis[i] = SafetyEmojiAlphabet[digest[i]%64]
	}
	return SafetyCode{Emojis: emojis, Digest: digest}, nil
}
